### DISCUSSION TITLE
Calendar Widget Theming and UX Improvements

### DISCUSSION THREAD

**Discussion by TriplEight** [pretalx/pretalx#1875] [Ideas] [answered]
First, cheers for the great conference management system. I'm currently self-hosting pretalx 2024.3.0 in production and have some observations and suggestions regarding the calendar widget theming and UX.

## Current Limitations

I recently attempted to embed the public agenda into a dark-themed site. While the Custom CSS feature works great for the outer parts of the public agenda, I discovered several challenges:

1. **Widget Theming**: The calendar widget uses Shadow DOM, making it impossible to style via Custom CSS. This isn't documented clearly, and it was quite surprising when my CSS changes affected everything except the widget itself.

2. **Persistence Issues**: Favorited talks ([stars](https://github.com/pretalx/pretalx/discussions/1786)) disappear upon page refresh, which affects user experience significantly.

## Feature Requests

Some improvements that would make the widget more flexible and user-friendly:

### Theming & Customization
- Support for dark/light themes in the [widget](https://github.com/pretalx/pretalx/issues/1610)
- Option to customize the widget's appearance from the admin UI
- Ability to color agenda card items based on [tags](https://github.com/pretalx/pretalx/issues/1710) and/or room colors
- Option to add custom text in the header (e.g., for app download instructions or legend). Actually, the docs example [has](https://docs.pretalx.org/user/event/widget/#2024-10-28) events colored after room's color, and I have no such option in rooms settings.

### UX Improvements
- **Favorite Button Visibility**: Currently, favorite stars only appear on cursor hover, making them inaccessible on touch devices
- **View Switching**: Add more options to switch between grid and line views (currently limited to horizontal layout)

## Questions

1. Are there any existing forks that implement some of these features?
2. Which files would need to be modified to implement these changes? (Ideally in a production-safe way)
3. Is there any documentation about the widget's architecture that could help in understanding how to extend it?

I believe these improvements would make pretalx even more versatile for various conference setups and user preferences. Happy to provide more details or clarification if needed.

**Comment by rixx** [comment, marked as answer]
> Persistence Issues: Favorited talks (https://github.com/pretalx/pretalx/discussions/1786) disappear upon page refresh, which affects user experience significantly.

If this is the case, this is a bug. However, I was not able to replicate the problem – when you are logged in, your starred talks are saved both to the server and locally, and if you are not logged in, they are stored just locally. I can reproduce this behaviour with both current Chrome and Firefox, logged-in and not, and the local saving has been around forever.

> Ability to color agenda card items based on https://github.com/pretalx/pretalx/issues/1710 and/or room colors
> […]
> Actually, the docs example [has](https://docs.pretalx.org/user/event/widget/#2024-10-28) events colored after room's color, and I have no such option in rooms settings.

What you are seeing in our demo events are sessions coloured by tracks, which is (and will remain, for the foreseeable future) our only separate colour option, because otherwise we’d have to resolve conflicts, when e.g. both a room and a track and a tag want to change a proposal’s colour. We do plan to display tags as colourful badges in the near future, though.

> Support for dark/light themes in the https://github.com/pretalx/pretalx/issues/1610

Happy to merge a pull request to that extent.

> Option to customize the widget's appearance from the admin UI

Again, happy to merge a PR assuming that the changes are something beneficial to everybody and not just driving up the maintenance burden.

> Favorite Button Visibility: Currently, favorite stars only appear on cursor hover, making them inaccessible on touch devices

Happy to merge a PR for this, as it should be a fairly minor change.

> Option to add custom text in the header (e.g., for app download instructions or legend).

Completely separate issue, as it would belong in pretalx-core rather than within the embeddable widget. I’d be open to discussing a PR for this.

> Are there any existing forks that implement some of these features?

Not to my knowledge – I don’t think there are many forks that plan to run alongside pretalx long-term, as I try to get features into core as much as I can, and open plugin APIs for everything else. That’s because I think people maintaining plugins is more beneficial to everybody involved than them maintaining forks. But of course, forks are also the sign of a healthy ecosystem, so if you want to start one up instead of contributing, that’s fine too!

> Which files would need to be modified to implement these changes? (Ideally in a production-safe way)

Some files live in this repository, in the `agenda` module, and everything to do with the embeddable widget lives over in pretalx/schedule.

> Is there any documentation about the widget's architecture that could help in understanding how to extend it?

There is not, though, again, I’m happy to merge contributions adding this, if they make sense and benefit everybody.

**Comment by TriplEight** [comment]
> If this is the case, this is a bug. However, I was not able to replicate the problem – when you are logged in, your starred talks are saved both to the server and locally...

Yep, it's not the case when user is logged in. This is why I mentioned the issue.

> sessions coloured by tracks

Thanks for the tip, this issue is solved then! Re conflicts: you can assign a precedence order like default < track < room < tag < custom session color. Maybe with less steps tho. This will prevent any conflict. And given docs coverage everyone will get it.

> Support for dark/light themes in the widget

AFAIU you're using fullcalendar.io. Upon checking their website they feature theming: https://fullcalendar.io/demos, https://fullcalendar.io/docs/theming. So this might be easier than appeared.

(and re all the PR requests: I'm more from the DevOps side of things, so would be amused if _PR barter_ would work :1st_place_medal:. I'm doing my part [here](https://github.com/pretalx/pretalx-docker/pull/64#issuecomment-2424799098) and [there](https://github.com/TriplEight/pretalx-docker/pulls).)

**Comment by rixx** [comment]
> Yep, it's not the case when user is logged in. This is why I mentioned the issue.

Doesn’t help me reproduce this, and for me, in Firefox and Chrome (current, plugins or no) the persistent stars work even in anonymous mode.

> Re conflicts: you can assign a precedence order like default < track < room < tag < custom session color. Maybe with less steps tho. This will prevent any conflict. And given docs coverage everyone will get it.

I wish I could have your confidence in users reading documentation and in more settings being a good idea. Alas, that has not been my experience. This is probably in fact a very good reason to start a fork!

> AFAIU you're using fullcalendar.io.

We are using fullcalendar for the availability widget, which I did not understand to be your primary reason to raise this feature request. Our schedule widget does not use external calendar applications, as I am not aware of any that would meet our requirements.
